File: scripts/check-duplicate-tax-ids.py
import json
import os
from collections import defaultdict


def find_duplicate_tax_ids(directory):
    tax_id_map = defaultdict(list)

    # Iterate over all files in the specified directory
    count = 0
    for filename in os.listdir(directory):
        if filename.endswith(".json"):
            print(f"Reading {filename}")
            filepath = os.path.join(directory, filename)
            with open(filepath, "r") as file:
                try:
                    # import pdb

                    # pdb.set_trace()
                    data = json.load(file)
                    users = data["Items"]
                    for user in users:
                        count += 1
                        user_id = user.get("userId", None)

                        print(f"\tProcessing user #{count}, userId: {user_id}")

                        businesses = user.get("data", {}).get("businesses", {})

                        # Iterate over each business and extract the encrypted tax ID
                        for business_id, business_data in businesses.items():
                            encrypted_tax_id = business_data.get("profileData", {}).get(
                                "encryptedTaxId", None
                            )
                            if encrypted_tax_id:
                                tax_id_map[encrypted_tax_id].append(
                                    (user_id, business_id)
                                )
                except json.JSONDecodeError:
                    print(f"Error decoding JSON from file: {filename}")

    # Check for duplicates and print them
    for encrypted_tax_id, entries in tax_id_map.items():
        if len(entries) > 1:
            print("\n")
            print(f"Duplicate encryptedTaxId found: {encrypted_tax_id}")
            for user_id, business_id in entries:
                print(f"  UserId: {user_id}, BusinessId: {business_id}")


# Specify the directory containing the JSON files
json_directory = "/Users/jhechter/Desktop/converted_data"
find_duplicate_tax_ids(json_directory)

File: scripts/checkDuplicateTaxIds.ts
import * as fs from "fs";
import * as path from "path";
import {AWSEncryptionDecryptionFactory} from "@businessnjgovnavigator/api/src/client/AwsEncryptionDecryptionFactory";

interface BusinessProfile {
  profileData?: {
    encryptedTaxId?: string;
  };
}

interface User {
  userId?: string;
  data?: {
    businesses?: Record<string, BusinessProfile>;
  };
}

interface FileData {
  Items: User[];
}

type TaxIdMap = Map<string, Array<{ userId: string | undefined; businessId: string }>>;

// Specify the directory containing the JSON files
const jsonDirectory = "/Users/jhechter/Desktop/converted_data";
const AWS_CRYPTO_KEY = "";
const AWS_CRYPTO_CONTEXT_STAGE = "";
const AWS_CRYPTO_CONTEXT_PURPOSE = "";
const AWS_CRYPTO_CONTEXT_ORIGIN =  "";

async function findDuplicateTaxIds(directory: string): Promise<void> {
  const taxIdMap: TaxIdMap = new Map();
  let count = 0;

  const AWSEncryptionDecryptionClient = AWSEncryptionDecryptionFactory(AWS_CRYPTO_KEY, {
    stage: AWS_CRYPTO_CONTEXT_STAGE,
    purpose: AWS_CRYPTO_CONTEXT_PURPOSE,
    origin: AWS_CRYPTO_CONTEXT_ORIGIN,
  });

  // Iterate over all files in the specified directory
  const files = fs.readdirSync(directory);
  files.forEach((filename) => {
    if (filename.endsWith(".json")) {
      console.log(`Reading ${filename}`);
      const filepath = path.join(directory, filename);

      try {
        const fileContent = fs.readFileSync(filepath, "utf-8");
        const data: FileData = JSON.parse(fileContent);

        data.Items.forEach((user) => {
          count++;
          const userId = user.userId;
          console.log(`\tProcessing user #${count}, userId: ${userId}`);

          const businesses = user.data?.businesses || {};

          // Iterate over each business and extract the encrypted tax ID
          Object.entries(businesses).forEach(async ([businessId, businessData]) => {
            const encryptedTaxId = businessData.profileData?.encryptedTaxId;
            let plainTextTaxId = await AWSEncryptionDecryptionClient.decryptValue(encryptedTaxId);
            if (plainTextTaxId) {
              if (!taxIdMap.has(plainTextTaxId)) {
                taxIdMap.set(plainTextTaxId, []);
              }
              taxIdMap.get(plainTextTaxId)?.push({ userId, businessId });
            }
          });
        });
      } catch (error) {
        console.error(`Error reading or parsing file: ${filename}`, error);
      }
    }
  });

  // Check for duplicates and print them
  taxIdMap.forEach((entries, encryptedTaxId) => {
    if (entries.length > 1) {
      console.log("\n");
      console.log(`Duplicate encryptedTaxId found: ${encryptedTaxId}`);
      entries.forEach(({ userId, businessId }) => {
        console.log(`  UserId: ${userId}, BusinessId: ${businessId}`);
      });
    }
  });

}
findDuplicateTaxIds(jsonDirectory).then(()=> {});


